File: codigo/bogo.go
package main

import (
	"fmt"
	"math/rand"
	"time"
)

// Problema 6c: iteraciones de bogo-sort ~ Geom(1/n!).
//
// Cada iteración genera una permutación uniforme. Hay n! permutaciones y solo
// una está ordenada, entonces p = 1/n!. Repetir hasta el primer éxito es una
// distribución geométrica.
func isSorted(arr []int) bool {
	for i := 0; i < len(arr)-1; i++ {
		if arr[i] > arr[i+1] {
			return false
		}
	}
	return true
}

func shuffle(arr []int) []int {
	copyArr := append([]int(nil), arr...)
	rand.Shuffle(len(copyArr), func(i, j int) {
		copyArr[i], copyArr[j] = copyArr[j], copyArr[i]
	})
	return copyArr
}

func bogoSort(arr []int) (iterations int, swaps int) {
	current := append([]int(nil), arr...)

	for {
		current = shuffle(current)
		iterations++
		swaps += len(current) - 1 // rand.Shuffle conceptualmente usa swaps; modelo similar a Fisher-Yates
		if isSorted(current) {
			break
		}
	}

	return iterations, swaps
}

func factorial(n int) int {
	result := 1
	for i := 2; i <= n; i++ {
		result *= i
	}
	return result
}

func simulate(n, trials int) {
	nFact := factorial(n)
	arr := make([]int, n)
	for i := range arr {
		arr[i] = i + 1
	}

	totalIterations := 0
	totalSwaps := 0
	for i := 0; i < trials; i++ {
		iterations, swaps := bogoSort(arr)
		totalIterations += iterations
		totalSwaps += swaps
	}

	fmt.Printf("\nn = %d (n! = %d, %d ensayos)\n", n, nFact, trials)
	fmt.Printf("  E[iteraciones] teórico  = %d\n", nFact)
	fmt.Printf("  E[iteraciones] simulado = %.2f\n", float64(totalIterations)/float64(trials))
	fmt.Printf("  E[swaps] teórico        = %d\n", (n-1)*nFact)
	fmt.Printf("  E[swaps] simulado       = %.2f\n", float64(totalSwaps)/float64(trials))
}

func main() {
	rand.Seed(time.Now().UnixNano())
	fmt.Println("Bogo-sort — verificando I ~ Geom(1/n!)")
	simulate(3, 5_000)
	simulate(4, 2_000)
	simulate(5, 500)
}

File: codigo/bogo.js
// ═══════════════════════════════════════════════════════════════
// PROBLEMA 6c: Iteraciones de bogo-sort ~ Geom(1/n!)
// ADA M5 · UVG
// ═══════════════════════════════════════════════════════════════
//
// ARGUMENTO:
//   Cada iteración de bogo-sort hace un shuffle uniforme aleatorio.
//   Hay n! permutaciones posibles y solo 1 está ordenada.
//   => P(éxito en una iteración) = 1/n!
//
//   Las iteraciones son INDEPENDIENTES: el shuffle borra la memoria
//   del pasado — no importa cuántas veces fallamos antes.
//
//   Experimentos de Bernoulli independientes con P(éxito) = p = 1/n!
//   contando intentos hasta el primer éxito = definición exacta de
//   distribución Geométrica.
//
//   I ~ Geom(1/n!)  =>  E[I] = n!
// ═══════════════════════════════════════════════════════════════

function isSorted(arr) {
  for (let i = 0; i < arr.length - 1; i++) {
    if (arr[i] > arr[i + 1]) return false;
  }
  return true;
}

// Fisher-Yates shuffle — produce permutación uniforme
function shuffle(arr) {
  const a = [...arr];
  for (let i = a.length - 1; i > 0; i--) {
    const j = Math.floor(Math.random() * (i + 1));
    [a[i], a[j]] = [a[j], a[i]];
  }
  return a;
}

function bogoSort(arr) {
  let current = [...arr];
  let iterations = 0;
  let swaps = 0;

  while (true) {
    current = shuffle(current);
    iterations++;
    swaps += current.length - 1; // Fisher-Yates hace n-1 swaps
    if (isSorted(current)) break;
  }

  return { iterations, swaps };
}

// ── Simulación ───────────────────────────────────────────────────
function factorial(n) {
  let f = 1;
  for (let i = 2; i <= n; i++) f *= i;
  return f;
}

function simulate(n, trials) {
  const nFact = factorial(n);
  const arr = Array.from({ length: n }, (_, i) => i + 1);

  let totalIter = 0;
  let totalSwaps = 0;

  for (let t = 0; t < trials; t++) {
    // Barajar primero para empezar desde permutación aleatoria
    const start = shuffle(arr);
    const { iterations, swaps } = bogoSort(start);
    totalIter += iterations;
    totalSwaps += swaps;
  }

  const avgIter = totalIter / trials;
  const avgSwaps = totalSwaps / trials;

  console.log(`\nn = ${n}  (n! = ${nFact}, ${trials} ensayos)`);
  console.log(`  E[iteraciones] teórico  = ${nFact}`);
  console.log(`  E[iteraciones] simulado = ${avgIter.toFixed(2)}`);
  console.log(`  E[swaps] teórico        = ${(n - 1) * nFact}`);
  console.log(`  E[swaps] simulado       = ${avgSwaps.toFixed(2)}`);
}

console.log("Bogo-sort — verificando I ~ Geom(1/n!)");
simulate(3, 5000); // n=3: n!=6, rápido
simulate(4, 2000); // n=4: n!=24
simulate(5, 500);  // n=5: n!=120, más lento
